File: packages/dotnet/src/SlyBrowser/SlyWebDriverSession.cs
using System.Diagnostics;
using OpenQA.Selenium.Remote;

namespace SlyBrowser;

public sealed class SlyWebDriverSession : IDisposable, IAsyncDisposable
{
    private readonly Process _driverProcess;
    private readonly Uri _driverServiceUrl;
    private readonly LaunchPlan _launchPlan;
    private readonly string _driverLicenseFile;
    private readonly string? _driverRuntimeFile;
    private readonly List<Func<ValueTask>> _closeCallbacks = [];
    private bool _disposed;

    internal SlyWebDriverSession(
        RemoteWebDriver driver,
        Process driverProcess,
        Uri driverServiceUrl,
        LaunchPlan launchPlan,
        string driverLicenseFile,
        string? driverRuntimeFile,
        string browserVersion,
        string driverVersion)
    {
        Driver = driver;
        _driverProcess = driverProcess;
        _driverServiceUrl = driverServiceUrl;
        _launchPlan = launchPlan;
        _driverLicenseFile = driverLicenseFile;
        _driverRuntimeFile = driverRuntimeFile;
        BrowserVersion = browserVersion;
        DriverVersion = driverVersion;
    }

    public RemoteWebDriver Driver { get; }
    public string BrowserVersion { get; }
    public string DriverVersion { get; }
    public LicenseRuntimeMetadata? LicenseRuntime { get; internal set; }

    internal void AddCloseCallback(Func<ValueTask> callback) => _closeCallbacks.Add(callback);

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        try { Driver.Quit(); }
        finally
        {
            RequestShutdown(_driverServiceUrl);
            StopProcess(_driverProcess);
            _launchPlan.Dispose();
            TryDelete(_driverRuntimeFile);
            TryDelete(_driverLicenseFile);
            Driver.Dispose();
            _driverProcess.Dispose();
            foreach (Func<ValueTask> callback in _closeCallbacks)
            {
                try { callback().AsTask().GetAwaiter().GetResult(); }
                catch { /* release is best effort on close; explicit service state remains authoritative */ }
            }
            GC.SuppressFinalize(this);
        }
    }

    public ValueTask DisposeAsync()
    {
        Dispose();
        return ValueTask.CompletedTask;
    }

    internal static void StopProcess(Process? process)
    {
        if (process is null) return;
        try
        {
            if (!process.HasExited)
            {
                process.Kill(entireProcessTree: true);
                process.WaitForExit(3_000);
            }
        }
        catch (InvalidOperationException) { }
    }

    internal static void RequestShutdown(Uri serviceUrl)
    {
        try
        {
            using HttpClient client = new() { Timeout = TimeSpan.FromSeconds(1) };
            client.GetAsync(new Uri(serviceUrl, "shutdown")).GetAwaiter().GetResult().Dispose();
        }
        catch { /* WebDriver shutdown is best effort; StopProcess remains the fallback */ }
    }

    internal static void TryDelete(string? path)
    {
        if (string.IsNullOrEmpty(path)) return;
        try { File.Delete(path); }
        catch (FileNotFoundException) { }
    }
}

File: packages/java/src/main/java/com/slybrowser/SlyWebDriverSession.java
package com.slybrowser;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeDriverService;

public final class SlyWebDriverSession implements AutoCloseable {
  private final ChromeDriver driver;
  private final ChromeDriverService service;
  private final LaunchPlan launchPlan;
  private final Path driverLicenseFile;
  private final Path driverRuntimeFile;
  private final String browserVersion;
  private final String driverVersion;
  private final List<Runnable> closeCallbacks = new ArrayList<>();
  LicenseRuntimeMetadata licenseRuntime;
  private boolean closed;

  SlyWebDriverSession(
      ChromeDriver driver,
      ChromeDriverService service,
      LaunchPlan launchPlan,
      Path driverLicenseFile,
      Path driverRuntimeFile,
      String browserVersion,
      String driverVersion) {
    this.driver = driver;
    this.service = service;
    this.launchPlan = launchPlan;
    this.driverLicenseFile = driverLicenseFile;
    this.driverRuntimeFile = driverRuntimeFile;
    this.browserVersion = browserVersion;
    this.driverVersion = driverVersion;
  }

  public ChromeDriver getDriver() { return driver; }
  public String getBrowserVersion() { return browserVersion; }
  public String getDriverVersion() { return driverVersion; }
  public LicenseRuntimeMetadata getLicenseRuntime() { return licenseRuntime; }
  void addCloseCallback(Runnable callback) { closeCallbacks.add(callback); }

  @Override
  public void close() {
    if (closed) return;
    closed = true;
    try { driver.quit(); }
    finally {
      try { service.close(); }
      finally {
        launchPlan.close();
        if (driverRuntimeFile != null) {
          try { Files.deleteIfExists(driverRuntimeFile); }
          catch (IOException ignored) { /* consumed one-time handoff or next cleanup pass */ }
        }
        try { Files.deleteIfExists(driverLicenseFile); }
        catch (IOException ignored) { /* consumed one-time handoff or next cleanup pass */ }
        for (Runnable callback : closeCallbacks) {
          try { callback.run(); }
          catch (RuntimeException ignored) { /* release is best effort on close */ }
        }
      }
    }
  }
}
